contracts/linear/src/validator_pool.rs

```diff
@@ -230,27 +255,70 @@ fn min3(x: u128, y: u128, z: u128) -> u128 {
 
 #[near_bindgen]
 impl LiquidStakingContract {
-    pub fn add_validator(&mut self, validator_id: AccountId, weight: u16) -> Validator {
+    pub fn set_whitelist_contract_id(&mut self, account_id: AccountId) {
+        self.assert_owner();
+        self.whitelist_account_id = Some(account_id);
+    }
+
+    pub fn add_validator(&mut self, validator_id: AccountId, weight: u16) {
         self.assert_manager();
-        self.validator_pool.add_validator(&validator_id, weight)
+        self.add_whitelisted_validator(&validator_id, weight);
     }
 
     pub fn add_validators(
         &mut self,
         validator_ids: Vec<AccountId>,
         weights: Vec<u16>,
-    ) -> Vec<Validator> {
+    ) {
         self.assert_manager();
         require!(validator_ids.len() == weights.len(), ERR_BAD_VALIDATOR_LIST);
-        let mut results: Vec<Validator> = vec![];
         for i in 0..validator_ids.len() {
-            results.push(
-                self.validator_pool
-                    .add_validator(&validator_ids[i], weights[i]),
-            );
+            self.add_whitelisted_validator(&validator_ids[i], weights[i]);
+        }
+    }
+    
+    /// Add a new validator only if it's whitelisted
+    fn add_whitelisted_validator(
+        &mut self,
+        validator_id: &AccountId,
+        weight: u16
+    ) {
+        if let Some(whitelist_id) = &self.whitelist_account_id {
+            ext_whitelist::is_whitelisted(
+                validator_id.clone(),
+                whitelist_id.clone(),
+                NO_DEPOSIT,
+                GAS_EXT_WHITELIST
+            )
+            .then(ext_self_whitelist_cb::is_whitelisted_callback(
+                validator_id.clone(),
+                weight,
+                env::current_account_id(),
+                NO_DEPOSIT,
+                GAS_CB_WHITELIST
+            ));
+        } else {
+            self.validator_pool.add_validator(validator_id, weight);
```

Comment:
Does this mean if whitelist account ID is not set, we could add non-whitelisted validator?

This means owner could bypass whitelist check when needed. I suggest we forbid this.

Reply:
this is meant to make contract compatible with current behavior after migration has done but before owner has set the whitelist account id.
The owner can bypass check anyway I think, since he can change whitelist account to whatever contract he wants..

Reply:
yeah. another way is to hardcode the whitelist account in contract code (for mainnet and testnet), since it should seldom (if not never) change.